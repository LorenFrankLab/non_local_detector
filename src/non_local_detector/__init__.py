from non_local_detector.models import (
    ContFragClusterlessClassifier,  # noqa
    ContFragSortedSpikesClassifier,  # noqa
    MultiEnvironmentSortedSpikesClassifier,  # noqa
    MultiEnvironmentClusterlessClassifier,  # noqa
    NonLocalClusterlessDetector,  # noqa
    NonLocalSortedSpikesDetector,  # noqa
    ClusterlessDecoder,  # noqa
    SortedSpikesDecoder,  # noqa
)

try:
    from ._version import __version__
except ImportError:
    pass
