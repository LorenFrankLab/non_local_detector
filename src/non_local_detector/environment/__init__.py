from non_local_detector.environment.alignment import (
    get_2d_rotation_matrix,
    map_probabilities_to_nearest_target_bin,
)
from non_local_detector.environment.environment import Environment
from non_local_detector.environment.layout.factories import (
    get_layout_parameters,
    list_available_layouts,
)

__all__ = [
    "get_2d_rotation_matrix",
    "map_probabilities_to_nearest_target_bin",
    "Environment",
    "get_layout_parameters",
    "list_available_layouts",
]
