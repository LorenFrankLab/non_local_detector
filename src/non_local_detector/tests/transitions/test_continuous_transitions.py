import numpy as np
import pytest
from syrupy.assertion import SnapshotAssertion

from non_local_detector.continuous_state_transitions import (
    Discrete,
    EmpiricalMovement,
    Identity,
    RandomWalk,
    RandomWalkDirection1,
    RandomWalkDirection2,
    Uniform,
)
from non_local_detector.environment import Environment


@pytest.fixture
def make_env_1d():
    """Factory fixture for creating 1D environments with custom parameters."""

    def _make_env(n_bins=11, name="line"):
        env = Environment(
            environment_name=name,
            place_bin_size=1.0,
            position_range=((0.0, float(n_bins - 1)),),
        )
        pos = np.linspace(0.0, float(n_bins - 1), n_bins)[:, None]
        env = env.fit_place_grid(position=pos, infer_track_interior=False)
        return env

    return _make_env


def rows_sum_to_one(mat: np.ndarray, atol=1e-7) -> bool:
    return np.allclose(mat.sum(axis=1), 1.0, atol=atol)


def test_random_walk_row_stochastic_and_spread_var(make_env_1d):
    env = make_env_1d(n_bins=21)
    envs = (env,)
    rw_tight = RandomWalk(
        environment_name=env.environment_name, movement_var=0.1
    ).make_state_transition(envs)
    rw_wide = RandomWalk(
        environment_name=env.environment_name, movement_var=10.0
    ).make_state_transition(envs)

    assert rw_tight.shape == (env.place_bin_centers_.shape[0],) * 2
    assert rows_sum_to_one(rw_tight) and rows_sum_to_one(rw_wide)
    assert np.all(rw_tight >= 0.0) and np.all(rw_wide >= 0.0)

    # Off-diagonal mass increases with larger variance
    offdiag_tight = (1.0 - np.diag(rw_tight)).mean()
    offdiag_wide = (1.0 - np.diag(rw_wide)).mean()
    assert offdiag_wide > offdiag_tight


def test_random_walk_with_directional_constraints(make_env_1d):
    env = make_env_1d(n_bins=15)
    envs = (env,)
    base = RandomWalk(
        environment_name=env.environment_name, movement_var=3.0
    ).make_state_transition(envs)
    up = RandomWalkDirection1(
        environment_name=env.environment_name, movement_var=3.0
    ).make_state_transition(envs)
    down = RandomWalkDirection2(
        environment_name=env.environment_name, movement_var=3.0
    ).make_state_transition(envs)

    # Directional versions are masked upper/lower triangular
    assert np.allclose(
        up,
        np.triu(base) / np.clip(np.triu(base).sum(axis=1, keepdims=True), 1e-12, None),
    )
    assert np.allclose(
        down,
        np.tril(base) / np.clip(np.tril(base).sum(axis=1, keepdims=True), 1e-12, None),
    )


def test_identity_transition_is_identity_masked(make_env_1d):
    env = make_env_1d(n_bins=10)
    envs = (env,)
    tm = Identity(environment_name=env.environment_name).make_state_transition(envs)
    assert tm.shape == (env.place_bin_centers_.shape[0],) * 2
    assert rows_sum_to_one(tm)
    assert np.allclose(np.diag(tm), 1.0)
    assert np.allclose(tm - np.diag(np.diag(tm)), 0.0)


def test_uniform_transition_within_env(make_env_1d):
    env = make_env_1d(n_bins=12)
    envs = (env,)
    tm = Uniform(environment_name=env.environment_name).make_state_transition(envs)
    n = env.place_bin_centers_.shape[0]
    assert tm.shape == (n, n)
    assert rows_sum_to_one(tm)
    # Each row is uniform
    assert np.allclose(tm, np.ones((n, n)) / n)


def test_uniform_transition_between_envs(make_env_1d):
    env1 = make_env_1d(n_bins=8, name="env1")
    env2 = make_env_1d(n_bins=5, name="env2")
    envs = (env1, env2)
    tm = Uniform(
        environment_name=env1.environment_name, environment2_name=env2.environment_name
    ).make_state_transition(envs)
    assert tm.shape == (
        env1.place_bin_centers_.shape[0],
        env2.place_bin_centers_.shape[0],
    )
    assert rows_sum_to_one(tm)
    # Rows uniform over destination bins
    assert np.allclose(np.unique(tm, axis=1), np.ones((tm.shape[0], 1)) / tm.shape[1])


def test_empirical_movement_row_stochastic_and_locality(make_env_1d):
    # Create a simple path that marches right
    env = make_env_1d(n_bins=11)
    # Ensure histogram range set for 2D (pre- and post- positions)
    env.position_range = ((0.0, 10.0), (0.0, 10.0))
    envs = (env,)
    pos = np.linspace(0.0, 10.0, 51)[:, None]
    # Use empirical transition with provided positions
    tm = EmpiricalMovement(
        environment_name=env.environment_name, speedup=1, is_time_reversed=False
    ).make_state_transition(envs, position=pos)
    assert tm.shape == (env.place_bin_centers_.shape[0],) * 2
    assert rows_sum_to_one(tm, atol=1e-6)
    # Empirical movement should favor near-diagonal transitions on average
    nearer = np.mean(np.diag(tm))
    farther = np.mean(tm[:, 0] + tm[:, -1])
    assert nearer >= farther


def test_discrete_transition_is_ones():
    tm = Discrete().make_state_transition()
    assert tm.shape == (1, 1) and np.isclose(tm[0, 0], 1.0)


# ============================================================================
# SNAPSHOT TESTS
# ============================================================================


def serialize_transition_matrix_summary(trans_mat: np.ndarray) -> dict:
    """Serialize transition matrix to summary for snapshot comparison.

    Parameters
    ----------
    trans_mat : np.ndarray
        Transition matrix

    Returns
    -------
    summary : dict
        Summary statistics suitable for snapshot comparison
    """
    return {
        "shape": trans_mat.shape,
        "dtype": str(trans_mat.dtype),
        "diagonal_mean": float(np.mean(np.diag(trans_mat)))
        if trans_mat.ndim == 2 and trans_mat.shape[0] == trans_mat.shape[1]
        else None,
        "diagonal_std": float(np.std(np.diag(trans_mat)))
        if trans_mat.ndim == 2 and trans_mat.shape[0] == trans_mat.shape[1]
        else None,
        "row_sums": np.sum(trans_mat, axis=-1).tolist()
        if trans_mat.size <= 500
        else {
            "mean": float(np.mean(np.sum(trans_mat, axis=-1))),
            "min": float(np.min(np.sum(trans_mat, axis=-1))),
            "max": float(np.max(np.sum(trans_mat, axis=-1))),
        },
        "mean": float(np.mean(trans_mat)),
        "std": float(np.std(trans_mat)),
        "min": float(np.min(trans_mat)),
        "max": float(np.max(trans_mat)),
        "values": trans_mat.tolist()
        if trans_mat.size <= 100
        else {
            "first_row": trans_mat[0].tolist() if trans_mat.shape[0] > 0 else [],
            "last_row": trans_mat[-1].tolist() if trans_mat.shape[0] > 0 else [],
            "diagonal_sample": np.diag(trans_mat)[:10].tolist()
            if trans_mat.ndim == 2 and trans_mat.shape[0] == trans_mat.shape[1]
            else None,
        },
    }


@pytest.mark.snapshot
def test_random_walk_snapshot(make_env_1d, snapshot: SnapshotAssertion):
    """Snapshot test for RandomWalk transition matrix."""
    env = make_env_1d(n_bins=15)
    envs = (env,)

    rw = RandomWalk(
        environment_name=env.environment_name,
        movement_var=2.0,
        movement_mean=0.0,
    ).make_state_transition(envs)

    assert serialize_transition_matrix_summary(rw) == snapshot


@pytest.mark.snapshot
def test_random_walk_with_mean_snapshot(make_env_1d, snapshot: SnapshotAssertion):
    """Snapshot test for RandomWalk with non-zero mean displacement."""
    env = make_env_1d(n_bins=12)
    envs = (env,)

    rw = RandomWalk(
        environment_name=env.environment_name,
        movement_var=1.5,
        movement_mean=0.5,
    ).make_state_transition(envs)

    assert serialize_transition_matrix_summary(rw) == snapshot


@pytest.mark.snapshot
def test_random_walk_direction1_snapshot(make_env_1d, snapshot: SnapshotAssertion):
    """Snapshot test for RandomWalkDirection1 (upper triangular)."""
    env = make_env_1d(n_bins=10)
    envs = (env,)

    rw_dir1 = RandomWalkDirection1(
        environment_name=env.environment_name, movement_var=2.5
    ).make_state_transition(envs)

    assert serialize_transition_matrix_summary(rw_dir1) == snapshot


@pytest.mark.snapshot
def test_random_walk_direction2_snapshot(make_env_1d, snapshot: SnapshotAssertion):
    """Snapshot test for RandomWalkDirection2 (lower triangular)."""
    env = make_env_1d(n_bins=10)
    envs = (env,)

    rw_dir2 = RandomWalkDirection2(
        environment_name=env.environment_name, movement_var=2.5
    ).make_state_transition(envs)

    assert serialize_transition_matrix_summary(rw_dir2) == snapshot


@pytest.mark.snapshot
def test_uniform_within_environment_snapshot(make_env_1d, snapshot: SnapshotAssertion):
    """Snapshot test for Uniform transition within same environment."""
    env = make_env_1d(n_bins=8)
    envs = (env,)

    uniform = Uniform(environment_name=env.environment_name).make_state_transition(envs)

    assert serialize_transition_matrix_summary(uniform) == snapshot


@pytest.mark.snapshot
def test_uniform_between_environments_snapshot(
    make_env_1d, snapshot: SnapshotAssertion
):
    """Snapshot test for Uniform transition between two environments."""
    env1 = make_env_1d(n_bins=6, name="env1")
    env2 = make_env_1d(n_bins=9, name="env2")
    envs = (env1, env2)

    uniform = Uniform(
        environment_name=env1.environment_name, environment2_name=env2.environment_name
    ).make_state_transition(envs)

    assert serialize_transition_matrix_summary(uniform) == snapshot


@pytest.mark.snapshot
def test_identity_snapshot(make_env_1d, snapshot: SnapshotAssertion):
    """Snapshot test for Identity transition matrix."""
    env = make_env_1d(n_bins=7)
    envs = (env,)

    identity = Identity(environment_name=env.environment_name).make_state_transition(
        envs
    )

    assert serialize_transition_matrix_summary(identity) == snapshot


@pytest.mark.snapshot
def test_empirical_movement_snapshot(make_env_1d, snapshot: SnapshotAssertion):
    """Snapshot test for EmpiricalMovement transition matrix."""
    env = make_env_1d(n_bins=11)
    env.position_range = ((0.0, 10.0), (0.0, 10.0))
    envs = (env,)

    # Create consistent position trajectory
    np.random.seed(42)
    pos = np.linspace(0.0, 10.0, 50)[:, None]

    empirical = EmpiricalMovement(
        environment_name=env.environment_name, speedup=1, is_time_reversed=False
    ).make_state_transition(envs, position=pos)

    assert serialize_transition_matrix_summary(empirical) == snapshot


@pytest.mark.snapshot
def test_empirical_movement_time_reversed_snapshot(
    make_env_1d, snapshot: SnapshotAssertion
):
    """Snapshot test for EmpiricalMovement with time reversal."""
    env = make_env_1d(n_bins=11)
    env.position_range = ((0.0, 10.0), (0.0, 10.0))
    envs = (env,)

    # Create consistent position trajectory
    np.random.seed(123)
    pos = np.linspace(0.0, 10.0, 50)[:, None]

    empirical = EmpiricalMovement(
        environment_name=env.environment_name, speedup=1, is_time_reversed=True
    ).make_state_transition(envs, position=pos)

    assert serialize_transition_matrix_summary(empirical) == snapshot


@pytest.mark.snapshot
def test_discrete_snapshot(snapshot: SnapshotAssertion):
    """Snapshot test for Discrete transition matrix."""
    discrete = Discrete().make_state_transition()

    assert serialize_transition_matrix_summary(discrete) == snapshot
