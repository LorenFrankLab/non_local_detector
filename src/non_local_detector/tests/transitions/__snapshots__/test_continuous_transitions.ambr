# serializer version: 1
# name: test_discrete_snapshot
  dict({
    'diagonal_mean': 1.0,
    'diagonal_std': 0.0,
    'dtype': 'float64',
    'max': 1.0,
    'mean': 1.0,
    'min': 1.0,
    'row_sums': list([
      1.0,
    ]),
    'shape': tuple(
      1,
      1,
    ),
    'std': 0.0,
    'values': list([
      list([
        1.0,
      ]),
    ]),
  })
# ---
# name: test_empirical_movement_snapshot
  dict({
    'diagonal_mean': 0.82,
    'diagonal_std': 0.059999999999999984,
    'dtype': 'float64',
    'max': 1.0,
    'mean': 0.1,
    'min': 0.0,
    'row_sums': list([
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
    ]),
    'shape': tuple(
      10,
      10,
    ),
    'std': 0.24738633753705966,
    'values': list([
      list([
        0.8,
        0.2,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.8,
        0.2,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.8,
        0.2,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.8,
        0.2,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.8,
        0.2,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.8,
        0.2,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.8,
        0.2,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.8,
        0.2,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.8,
        0.2,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        1.0,
      ]),
    ]),
  })
# ---
# name: test_empirical_movement_time_reversed_snapshot
  dict({
    'diagonal_mean': 0.8200000000000001,
    'diagonal_std': 0.059999999999999984,
    'dtype': 'float64',
    'max': 1.0,
    'mean': 0.1,
    'min': 0.0,
    'row_sums': list([
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
    ]),
    'shape': tuple(
      10,
      10,
    ),
    'std': 0.24738633753705966,
    'values': list([
      list([
        1.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.2,
        0.8,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.2,
        0.8,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.2,
        0.8,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.2,
        0.8,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.2,
        0.8,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.2,
        0.8,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.2,
        0.8,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.2,
        0.8,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.2,
        0.8,
      ]),
    ]),
  })
# ---
# name: test_identity_snapshot
  dict({
    'diagonal_mean': 1.0,
    'diagonal_std': 0.0,
    'dtype': 'float64',
    'max': 1.0,
    'mean': 0.16666666666666666,
    'min': 0.0,
    'row_sums': list([
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
    ]),
    'shape': tuple(
      6,
      6,
    ),
    'std': 0.372677996249965,
    'values': list([
      list([
        1.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        1.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        1.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        1.0,
        0.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        1.0,
        0.0,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        1.0,
      ]),
    ]),
  })
# ---
# name: test_random_walk_direction1_snapshot
  dict({
    'diagonal_mean': 0.4908711965316943,
    'diagonal_std': 0.1855857440602243,
    'dtype': 'float64',
    'max': 1.0,
    'mean': 0.1111111111111111,
    'min': 0.0,
    'row_sums': list([
      1.0000000000000002,
      1.0,
      1.0,
      0.9999999999999999,
      0.9999999999999999,
      0.9999999999999999,
      1.0,
      1.0,
      1.0,
    ]),
    'shape': tuple(
      9,
      9,
    ),
    'std': 0.18376628214549345,
    'values': list([
      list([
        0.40295550867748803,
        0.32991206707644105,
        0.18105958129938382,
        0.06660809758715265,
        0.01642535463891808,
        0.002715092860458436,
        0.0003008408641858187,
        2.2344527456173168e-05,
        1.1124685161081067e-06,
      ]),
      list([
        0.0,
        0.4029559569533035,
        0.329912434093637,
        0.1810597827226916,
        0.06660817168664655,
        0.016425372911628308,
        0.0027150958809171216,
        0.00030084119886218067,
        2.2344552313784116e-05,
      ]),
      list([
        0.0,
        0.0,
        0.4029649610249557,
        0.32991980600400167,
        0.18106382851287997,
        0.06660966004968015,
        0.016425739937433636,
        0.0027151565498746903,
        0.0003008479211742939,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.4030862286789077,
        0.3300190915616457,
        0.1811183175822111,
        0.06662970545805565,
        0.01643068307428,
        0.002715973644899722,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.40418398172095243,
        0.33091785573645266,
        0.1816115698194496,
        0.06681116281544751,
        0.0164754299076977,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.4109546360219759,
        0.3364611986311603,
        0.18465382090292431,
        0.06793034444393939,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.44090549839518783,
        0.3609828907373151,
        0.1981116108674971,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.5498339973124778,
        0.4501660026875221,
      ]),
      list([
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        1.0,
      ]),
    ]),
  })
# ---
# name: test_random_walk_direction2_snapshot
  dict({
    'diagonal_mean': 0.4908711965316943,
    'diagonal_std': 0.1855857440602243,
    'dtype': 'float64',
    'max': 1.0,
    'mean': 0.11111111111111113,
    'min': 0.0,
    'row_sums': list([
      1.0,
      1.0,
      0.9999999999999999,
      1.0,
      1.0,
      1.0,
      0.9999999999999999,
      0.9999999999999999,
      1.0,
    ]),
    'shape': tuple(
      9,
      9,
    ),
    'std': 0.18376628214549345,
    'values': list([
      list([
        1.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.45016600268752216,
        0.549833997312478,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.19811161086749707,
        0.36098289073731504,
        0.4409054983951878,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0679303444439394,
        0.18465382090292434,
        0.33646119863116036,
        0.410954636021976,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.016475429907697705,
        0.06681116281544752,
        0.18161156981944965,
        0.3309178557364527,
        0.4041839817209525,
        0.0,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.0027159736448997226,
        0.016430683074280004,
        0.06662970545805565,
        0.18111831758221114,
        0.33001909156164577,
        0.40308622867890775,
        0.0,
        0.0,
        0.0,
      ]),
      list([
        0.00030084792117429383,
        0.00271515654987469,
        0.016425739937433632,
        0.06660966004968014,
        0.18106382851287994,
        0.3299198060040016,
        0.40296496102495566,
        0.0,
        0.0,
      ]),
      list([
        2.234455231378412e-05,
        0.0003008411988621807,
        0.002715095880917121,
        0.016425372911628308,
        0.06660817168664655,
        0.1810597827226916,
        0.329912434093637,
        0.40295595695330344,
        0.0,
      ]),
      list([
        1.1124685161081065e-06,
        2.2344527456173165e-05,
        0.00030084086418581864,
        0.0027150928604584355,
        0.016425354638918078,
        0.06660809758715264,
        0.1810595812993838,
        0.329912067076441,
        0.402955508677488,
      ]),
    ]),
  })
# ---
# name: test_random_walk_snapshot
  dict({
    'diagonal_mean': 0.3129204443637498,
    'diagonal_std': 0.054106427249762654,
    'dtype': 'float64',
    'max': 0.44005294083377094,
    'mean': 0.07142857142857142,
    'min': 1.970439329244888e-19,
    'row_sums': list([
      1.0,
      0.9999999999999999,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0000000000000002,
      1.0,
      1.0000000000000002,
      1.0,
      1.0000000000000002,
      0.9999999999999999,
      1.0,
    ]),
    'shape': tuple(
      14,
      14,
    ),
    'std': 0.10627491111297678,
    'values': dict({
      'diagonal_sample': list([
        0.44005294083377094,
        0.32773403729226874,
        0.2924717130189193,
        0.28372550773807037,
        0.2822587181488455,
        0.2821050027841565,
        0.28209519073021677,
        0.2820951907302168,
        0.28210500278415657,
        0.2822587181488455,
      ]),
      'first_row': list([
        0.44005294083377094,
        0.34271357491421545,
        0.16188642995977737,
        0.04638123873004741,
        0.008059850756236855,
        0.0008495020197917202,
        5.430684721606274e-05,
        2.1057049806411945e-06,
        4.95214345824521e-08,
        7.063853263933016e-10,
        6.111430539911731e-12,
        3.2069844203505836e-14,
        1.0207128427797449e-16,
        1.970439329244888e-19,
      ]),
      'last_row': list([
        1.970439329244888e-19,
        1.0207128427797449e-16,
        3.2069844203505836e-14,
        6.111430539911731e-12,
        7.063853263933016e-10,
        4.95214345824521e-08,
        2.1057049806411945e-06,
        5.430684721606274e-05,
        0.0008495020197917202,
        0.008059850756236855,
        0.04638123873004741,
        0.16188642995977737,
        0.34271357491421545,
        0.44005294083377094,
      ]),
    }),
  })
# ---
# name: test_random_walk_with_mean_snapshot
  dict({
    'diagonal_mean': 0.3435826706663979,
    'diagonal_std': 0.0854536738856436,
    'dtype': 'float64',
    'max': 0.599381349400351,
    'mean': 0.09090909090909091,
    'min': 7.137928971059536e-17,
    'row_sums': list([
      1.0,
      1.0,
      1.0000000000000002,
      1.0,
      1.0,
      1.0,
      1.0000000000000002,
      1.0,
      1.0,
      0.9999999999999999,
      1.0,
    ]),
    'shape': tuple(
      11,
      11,
    ),
    'std': 0.1283208141140035,
    'values': dict({
      'diagonal_sample': list([
        0.599381349400351,
        0.37475824613377823,
        0.3142871118182962,
        0.3014645767284889,
        0.29980917617455083,
        0.2996949031098989,
        0.2996949031098989,
        0.29980917617455083,
        0.3014645767284889,
        0.31428711181829616,
      ]),
      'first_row': list([
        0.599381349400351,
        0.30773264561099567,
        0.08111744468783962,
        0.010978052352259043,
        0.0007627929651395526,
        2.7211871163719292e-05,
        4.98402805721443e-07,
        4.6867619594422976e-09,
        2.2627452373178817e-11,
        5.608784682243135e-14,
        7.137928971059536e-17,
      ]),
      'last_row': list([
        3.5068463717836976e-14,
        1.4147628007330728e-11,
        2.930359267475193e-09,
        3.1162224438112085e-07,
        1.7013998052385516e-05,
        0.0004769300113606697,
        0.006863936706761646,
        0.05071801338577108,
        0.19240729906371143,
        0.37475824613377823,
        0.37475824613377823,
      ]),
    }),
  })
# ---
# name: test_uniform_between_environments_snapshot
  dict({
    'diagonal_mean': None,
    'diagonal_std': None,
    'dtype': 'float64',
    'max': 0.125,
    'mean': 0.125,
    'min': 0.125,
    'row_sums': list([
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
    ]),
    'shape': tuple(
      5,
      8,
    ),
    'std': 0.0,
    'values': list([
      list([
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
      ]),
      list([
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
      ]),
      list([
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
      ]),
      list([
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
      ]),
      list([
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
        0.125,
      ]),
    ]),
  })
# ---
# name: test_uniform_within_environment_snapshot
  dict({
    'diagonal_mean': 0.14285714285714282,
    'diagonal_std': 2.7755575615628914e-17,
    'dtype': 'float64',
    'max': 0.14285714285714285,
    'mean': 0.14285714285714285,
    'min': 0.14285714285714285,
    'row_sums': list([
      0.9999999999999998,
      0.9999999999999998,
      0.9999999999999998,
      0.9999999999999998,
      0.9999999999999998,
      0.9999999999999998,
      0.9999999999999998,
    ]),
    'shape': tuple(
      7,
      7,
    ),
    'std': 0.0,
    'values': list([
      list([
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
      ]),
      list([
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
      ]),
      list([
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
      ]),
      list([
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
      ]),
      list([
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
      ]),
      list([
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
      ]),
      list([
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
        0.14285714285714285,
      ]),
    ]),
  })
# ---
